Minimal changes to get pmxbot up and running again

chrisBrookes93 opened the pull request:
Minimal changes to get pmxbot up and running again. Our main use case is webhooks as we've lost observability on applications as a result. This is now working as well as the common integrations (e.g. thanks and motivate)

Key changes:
- Changed the dependency to be the new name: `slack_sdk`
- Updated to use the v2 of the RTMClient
- `self.slack.server` is no longer a thing in the new client, so changed to use `self.slacker` for retrieving info on `users`/`channels` (or as they're referred to as now: `conversations`)
- Now using the web client for posting messages to slack. For this we need the channel ID rather than the channel name, so there is logic implemented to resolve this where necessary.
- Kept it simple by spawning a thread to run the scheduler rather than an asyncio loop as suggested by jaraco here: https://github.com/pmxbot/pmxbot/issues/85#issuecomment-489086969

Talking to PmxBot directly:
![image](https://user-images.githubusercontent.com/1598192/198656272-1a3eeaf8-4a2a-4e45-b5c5-3bc776fa8f5e.png)

Webhooks and scheduled messages:

![image](https://user-images.githubusercontent.com/1598192/198656597-a764c05b-e0c2-440d-baaf-92c86c94b452.png)


Responding in threads, tagging users and channels:

![image](https://user-images.githubusercontent.com/1598192/198656908-38e26f95-97b9-49f9-b570-11f99d51e62c.png)


Help docs:

![image](https://user-images.githubusercontent.com/1598192/198657514-1fd87f57-cad5-415d-9494-e3d6d5879475.png)



Some other random commands:

![image](https://user-images.githubusercontent.com/1598192/198658379-dc7bf05f-e3e4-4eec-9817-c86d629d5afa.png)


![image](https://user-images.githubusercontent.com/1598192/198660269-e496b381-80a6-4d43-9f85-f36cadacc125.png)

jaraco commented on setup.cfg:
```diff
@@ -95,7 +95,7 @@ viewer =
 
 slack =
 	# pinned due to #85
```
This comment becomes stale

chrisBrookes93 replied:
Agreed, removed

jaraco commented on pmxbot/slack.py:
```diff
@@ -86,27 +92,25 @@ def handle_message(self, msg):
-        channel = core.AugmentableMessage(channel, thread=msg.get('thread_ts'))
+        channel_name = (
+            self.slacker.conversations.info(msg['channel'])
+            .body.get('channel', {})
+            .get('name')
```
This code looks potentially overly-conservative. Will it work the same with `.body['channel']['name']`? I would be surprised if passing `None` for the `channel_name` would work as expected.

chrisBrookes93 replied:
There is a valid case where a message doesn't have a name, which is when you're talking directly to pmxbot (an IM)
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/1598192/198834190-59249c87-0483-43e9-aced-93d3704488c6.png">

As also seen in the example from the docs: https://api.slack.com/methods/conversations.list#:~:text=%22dGVhbTpDMDYxRkE1UEI%3D%22%0A%20%20%20%20%7D%0A%7D-,Alternate%20response,-Example%20response%20when

In this case I think we'd want to continue with a channel of `None` so that the actions still function, since we'll only need the channel_id.

So `name` isn't always present, but I think its fair to assume that `channel` is always present in the body if the request is successful so I will change to:

`self.slacker.conversations.info(msg['channel']).body['channel'].get('name')`

... but will also adapt so that we're not using `slacker` as you suggested in another comment.

jaraco replied:
Thanks for the investigation and explanation.

jaraco commented on pmxbot/slack.py:
```diff
@@ -64,14 +65,19 @@ def get_email_username_map(self, ttl_hash=get_ttl_hash()):
-                self.handle_message(msg)
             self.scheduler.run_pending()
-            time.sleep(0.1)
+            time.sleep(1)
```
Why change the sleep time here? Setting to `1` would mean that scheduled events could happen up to 1 sec after they're scheduled, which could potentially create a user-visible delay.

chrisBrookes93 replied:
My thinking was that 0.1 was a little low and would mean a lot of loop spinning. I think in reality increasing it to 1 would have little noticeable impact. From what I understand any pmxbot actions generated from direct user interaction (e.g. `!thank`) do not get pushed onto the scheduler queue, they're dealt with when the message event is received. This would only be a slight delay in the actions of other plugins (e.g. standup notifications and webhooks) by up to one second.

However, sticking with the philosophy of minimal changes here, I will change back to `0.1`

jaraco commented on pmxbot/slack.py:
```diff
@@ -115,16 +119,33 @@ def transmit(self, channel, message):
 
-        target.send_message(message, thread=getattr(channel, 'thread', None))
+        channel_name = channel.strip('#')
+        cursor = None
+        while True:
+            resp = self.slacker.conversations.list(cursor=cursor, exclude_archived=True)
+            if not resp.successful:
+                log.error('Failed calls to conversations.list')
+                return None
+            cursor = resp.body['response_metadata']['next_cursor']
+            chan_mapping = dict([(x['name'], x['id']) for x in resp.body['channels']])
+            if channel_name in chan_mapping:
+                return chan_mapping[channel_name]
+            if not cursor:
+                break
```
I'd like to refactor this bit to separate concerns (generating channel mappings, consuming channel mappings). I'll help with this.

jaraco replied:
Done in 6dfa674.

jaraco commented on pmxbot/slack.py:
```diff
@@ -64,14 +65,19 @@ def get_email_username_map(self, ttl_hash=get_ttl_hash()):
-        res = self.slack.rtm_connect()
-        assert res, "Error connecting"
         self.init_schedule(self.scheduler)
+        threading.Thread(target=self.run_scheduler_loop).start()
```
I wonder if `self.slack` has an event loop that could periodically invoke the scheduler rather than creating a thread. A thread may be fine, but it could also introduce threading issues (what if a message arrives at the same time as a scheduled event).

jaraco replied:
Oh, I see you answered the question in the description; "keep it simple". Sounds good. I've looked into the rtm API code and it uses threads to manage its concurrency, so I'm not too worried about concurrency concerns.

jaraco commented on pmxbot/slack.py:
```diff
@@ -64,14 +65,19 @@ def get_email_username_map(self, ttl_hash=get_ttl_hash()):
         self.init_schedule(self.scheduler)
+        threading.Thread(target=self.run_scheduler_loop).start()
+
+        @self.slack.on("message")
+        def handle_payload(client, event):
+            self.handle_message(event)
```
Did you consider instead using `self.slack.on("message")(self.handle_message)` and updating `handle_message` to accept `(client, event)`?

chrisBrookes93 replied:
agreed, actioned

chrisBrookes93 replied:
hmm, apparently not possible as slack validates the function signature:
```
    raise SlackClientError(error)
slack_sdk.errors.SlackClientError: The listener 'handle_message' must accept two args: client, event (actual: self, client, msg)
```

We can't really make `handle_message` static since it uses `self` all over

jaraco reviewed (approved):
Looks good. A few concerns to be addressed, but overall, it's nearly ready. I'll also want to see the tests passing, but the failures may be unrelated (I haven't investigated yet).

jaraco commented:
Tests were failing. Fixed in 6f8fa0b.

jaraco commented on pmxbot/slack.py:
```diff
@@ -115,16 +119,33 @@ def transmit(self, channel, message):
+        channel_name = channel.strip('#')
+        cursor = None
+        while True:
+            resp = self.slacker.conversations.list(cursor=cursor, exclude_archived=True)
```
Oh. I notice that `slacker` hasn't been updated in years and has been abandoned. It's probably not wise to add new behavior against a deprecated API. Is there any way to implement this behavior using `self.slack`?

chrisBrookes93 replied:
I agree, I will rework this and the other place that was changed to use `slacker`

chrisBrookes93 replied:
Use of `slacker` removed entirely

jaraco commented on pmxbot/slack.py:
```diff
@@ -115,16 +119,33 @@ def transmit(self, channel, message):
+        cursor = None
+        while True:
+            resp = self.slacker.conversations.list(cursor=cursor, exclude_archived=True)
+            if not resp.successful:
+                log.error('Failed calls to conversations.list')
+                return None
```
Is the right thing to do on an unsuccessful response to pretend like nothing happened? This `None` value will get cached. My instinct is this condition should trigger an Exception.

chrisBrookes93 replied:
I hadn't considered the caching of None, I agree with your approach, I will action.

chrisBrookes93 replied:
Using `self.slack.web_client.<endpoint>` to talk to slack does not return a `successful` field. Instead it will raise an exception, so I think this gives us what we want.

chrisBrookes93 commented:
@jaraco Thanks for the review and contributions, I particularly like your implementation for handling paginated data from slack!

I believe I've addressed all of your comments, please take another look when you have capacity.

I'd like to do some final testing tomorrow, so if you review and approve before then, don't merge just yet!

chrisBrookes93 commented on pmxbot/slack.py:
```diff
@@ -4,6 +4,8 @@
 import logging
 import re
 import html
+import threading
```
TODO - tidy up git history before merge

chrisBrookes93 commented:
> @jaraco Thanks for the review and contributions, I particularly like your implementation for handling paginated data from slack!
> 
> I believe I've addressed all of your comments, please take another look when you have capacity.
> 
> I'd like to do some final testing tomorrow, so if you review and approve before then, don't merge just yet!

... All looks good to me!

jaraco commented:
Excellent work and great collaboration. Thanks!
